File: src/main/java/engineer/carrot/warren/warren/irc/handlers/RPL/TopicHandler.java
package engineer.carrot.warren.warren.irc.handlers.RPL;

import engineer.carrot.warren.warren.event.TopicNotifyEvent;
import engineer.carrot.warren.warren.irc.Channel;
import engineer.carrot.warren.warren.irc.handlers.MessageHandler;
import engineer.carrot.warren.warren.irc.messages.RPL.TopicMessage;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

public class TopicHandler extends MessageHandler<TopicMessage> {
    private final Logger LOGGER = LoggerFactory.getLogger(TopicHandler.class);

    @Override
    public void handleMessage(TopicMessage message) {
        Channel channel = this.botDelegate.getJoinedChannels().getChannel(message.forChannel);
        if (channel == null) {
            LOGGER.warn("Got a topic message for a channel we aren't in: {} '{}'", message.forChannel, message.contents);

            return;
        }

        this.eventSink.postEvent(new TopicNotifyEvent(channel, message.contents));
    }
}

File: src/main/kotlin/engineer/carrot/warren/warren/handler/TopicHandler.kt
package engineer.carrot.warren.warren.handler

import engineer.carrot.warren.kale.IKaleHandler
import engineer.carrot.warren.kale.irc.message.rfc1459.TopicMessage
import engineer.carrot.warren.warren.loggerFor
import engineer.carrot.warren.warren.state.CaseMappingState
import engineer.carrot.warren.warren.state.ChannelsState

class TopicHandler(val channelsState: ChannelsState, val caseMappingState: CaseMappingState) : IKaleHandler<TopicMessage> {
    private val LOGGER = loggerFor<TopicHandler>()

    override val messageType = TopicMessage::class.java

    override fun handle(message: TopicMessage) {
        val channel = channelsState.getJoined(message.channel, caseMappingState.mapping)
        val topic = message.topic

        if (channel == null) {
            LOGGER.warn("got a topic for a channel we don't think we're in, not doing anything: $message")
            return
        }

        LOGGER.debug("channel topic for ${channel.name}: $topic")
        channel.topic = topic
    }
}

